dev/set_matrix.py

```diff
@@ -0,0 +1,489 @@
+            version = dev_numeric
+
+        for ver_spec, packages in requirements.items():
+            op_and_ver_pairs = map(get_operator_and_version, ver_spec.split(","))
+            match_all = all(
+                comp_op(
+                    LooseVersion(version),
+                    LooseVersion(dev_numeric if req_ver == DEV_VERSION else req_ver),
+                )
+                for comp_op, req_ver in op_and_ver_pairs
+            )
+            if match_all:
+                return packages
+        else:
+            return []
+
+    raise TypeError("Invalid object type for `requirements`: '{}'".format(type(requirements)))
+
+
+def remove_comments(s):
+    """
+    Examples
+    --------
+    >>> code = '''
+    ... # comment 1
+    ...  # comment 2
+    ... echo foo
+    ... '''
+    >>> remove_comments(code)
+    'echo foo'
+    """
+    return "\n".join(l for l in s.strip().split("\n") if not l.strip().startswith("#"))
+
+
+def make_pip_install_command(packages):
+    """
+    Examples
+    --------
+    >>> make_pip_install_command(["foo", "bar"])
+    "pip install 'foo' 'bar'"
+    """
+    return "pip install " + " ".join("'{}'".format(x) for x in packages)
+
+
+def divider(title, length=None):
+    r"""
+    Examples
+    --------
+    >>> divider("1234", 20)
+    '\n======= 1234 ======='
+    """
+    length = shutil.get_terminal_size(fallback=(80, 24))[0] if length is None else length
+    rest = length - len(title) - 2
+    left = rest // 2 if rest % 2 else (rest + 1) // 2
+    return "\n{} {} {}".format("=" * left, title, "=" * (rest - left))
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description="Set a test matrix for the cross version tests")
+    parser.add_argument(
+        "--ref-versions-yaml",
+        required=False,
+        default=None,
+        help=(
+            "URL or local file path of the reference config which will be compared with the config "
+            "on the branch where this script is running in order to identify version YAML updates"
+        ),
+    )
+    parser.add_argument(
+        "--changed-files",
+        type=lambda x: [] if x.strip() == "" else x.strip().split("\n"),
+        required=False,
+        default=None,
+        help=("A string that represents a list of changed files"),
+    )
+
+    return parser.parse_args()
+
+
+class Hashabledict(dict):
+    def __hash__(self):
+        return hash(frozenset(self))
+
+
+def expand_config(config):
+    matrix = []
+    for flavor_key, cfgs in config.items():
+        flavor = flavor_key.split("-")[0]
+        package_info = cfgs.pop("package_info")
+
+        for key, cfg in cfgs.items():
+            # Released versions
+            versions = get_released_versions(package_info["pip_release"])
+            versions = filter_versions(
+                versions, cfg["minimum"], cfg["maximum"], cfg.get("unsupported"),
+            )
+            versions = select_latest_micro_versions(versions)
```

Comment:
@harupy We also want to explicitly include the minimum supported version in our testing procedure. For example, `scikit-learn` tests should run against `0.20.3` *and* `0.20.4`.